### PR TITLE
feat: migrate multi-agent tools to sdk native patterns and model support

### PR DESCRIPTION
**Pull request by cagataycali**
## Description

This PR represents a **major architectural refactor** that migrates all multi-agent tools from custom implementations to native Strands SDK patterns while introducing comprehensive model provider support. **These are breaking changes** - existing code using these tools will need to be updated.

### 🚨 Breaking Changes Summary

| Old Tool | New Tool | Breaking Change |
|----------|----------|-----------------|
| `agent_graph` | `graph` | Tool name changed, complete API redesign |
| `use_llm` | `use_agent` | Tool name changed, signature enhanced |
| `swarm` | `swarm` | API completely redesigned, agents parameter required |
| `think` | `think` | Enhanced with model switching (backward compatible) |

Also "workflow" tool is refactored to use @tool decorator. 

## Related Issues

Main tracking ticket:
https://github.com/strands-agents/sdk-python/issues/214

Related recent other issues:
https://github.com/strands-agents/tools/issues/136
https://github.com/strands-agents/tools/issues/128
https://github.com/strands-agents/tools/issues/74
https://github.com/strands-agents/tools/issues/20

Related other PRs:

https://github.com/strands-agents/tools/pull/137
https://github.com/strands-agents/tools/pull/129
https://github.com/strands-agents/tools/pull/117
https://github.com/strands-agents/tools/pull/71

## Documentation PR

[Will be created after this PR is merged]

## Type of Change

- [x] Breaking change  
- [x] New Tool (enhanced versions)
- [x] Bug fix
- [ ] Other (please describe):

### 1. **Graph Tool** (`agent_graph` → `graph`)

**Before:**
```python
result = agent.tool.agent_graph(
    action="create",
    graph_id="analysis_graph",
    topology={
        "type": "star",
        "nodes": [...],
        "edges": [...]
    }
)
```

**After:**
```python
result = agent.tool.graph(
    action="create", 
    graph_id="analysis_pipeline",
    topology={
        "nodes": [
            {
                "id": "researcher",
                "role": "researcher", 
                "system_prompt": "You research thoroughly.",
                "model_provider": "bedrock",
                "model_settings": {"model_id": "us.anthropic.claude-sonnet-4-20250514-v1:0"}
            }
        ],
        "edges": [{"from": "researcher", "to": "analyst"}],
        "entry_points": ["researcher"]
    }
)
```

**Changes:**
- Tool name changed from `agent_graph` to `graph`
- Migrated from custom message-passing to Strands SDK `GraphBuilder`
- Deterministic DAG execution instead of persistent queues
- Per-node model configuration support
- `entry_points` required instead of auto-detection

### 2. **Agent Creation** (`use_llm` → `use_agent`)

**Before:**
```python
result = agent.tool.use_llm(
    prompt="Calculate 2+2",
    system_prompt="You are a math helper"
)
```

**After:**
```python
result = agent.tool.use_agent(
    prompt="Calculate 2+2", 
    system_prompt="You are a math helper",
    model_provider="bedrock",  # NEW: Model switching support
    model_settings={"model_id": "us.anthropic.claude-sonnet-4-20250514-v1:0"},
    tools=["calculator"]  # NEW: Tool filtering support
)
```

**Changes:**
- Tool name changed from `use_llm` to `use_agent`
- Added `model_provider` parameter for model switching
- Added `model_settings` parameter for custom configuration
- Enhanced `tools` parameter for selective tool access

### 3. **Swarm Tool** - Complete Architecture Overhaul

**Before:**
```python
result = agent.tool.swarm(
    task="Analyze this problem",
    swarm_size=3,
    coordination_pattern="collaborative"
)
```

**After:**
```python
result = agent.tool.swarm(
    task="Develop comprehensive strategy",
    agents=[  # REQUIRED: Custom agent specifications
        {
            "name": "researcher",
            "system_prompt": "You are a research specialist.",
            "tools": ["retrieve", "calculator"],
            "model_provider": "bedrock"
        },
        {
            "name": "strategist", 
            "system_prompt": "You are a strategic planner.",
            "tools": ["file_write"],
            "model_provider": "anthropic"
        }
    ]
)
```

**Changes:**
- Removed `swarm_size` and `coordination_pattern` parameters
- **Required `agents` parameter** with custom agent specifications
- Migrated to Strands SDK `Swarm` pattern
- Per-agent model provider and tool configuration
- Autonomous coordination through injected tools

### 4. **Think Tool** - Enhanced (Backward Compatible)

**Enhanced Features (optional):**
```python
result = agent.tool.think(
    thought="How can AI be more creative?",
    cycle_count=3,
    system_prompt="You are a creative AI researcher.",
    model_provider="bedrock",  # NEW: Model switching
    model_settings={"model_id": "claude-sonnet-4"},  # NEW: Custom config
    thinking_system_prompt="Use design thinking methodology"  # NEW: Custom methodology
)
```

### **Model Provider Infrastructure**
- **Complete model utilities system** in `src/strands_tools/models/`
- **9+ providers supported**: Bedrock, Anthropic, LiteLLM, LlamaAPI, Ollama, OpenAI, Writer, Cohere, GitHub
- **Environment-based configuration** with fallback mechanisms
- **Per-task/agent model customization** across all multi-agent tools

### **SDK Native Patterns** 
- **GraphBuilder integration** for deterministic DAG execution
- **Swarm pattern** for autonomous agent coordination  
- **Output propagation model** for better context sharing
- **Enhanced error handling** and execution monitoring

### **Advanced Capabilities**
- **Custom agent team creation** with individual specifications
- **Mixed model provider teams** for optimized performance
- **Recursive thinking with model switching** for specialized cognitive tasks
- **Production-ready orchestration** with advanced task management

### Pre-commit Hooks Passed:
- [x] `hatch fmt --linter`
- [x] `hatch fmt --formatter` 
- [x] `hatch test --all`

### Additional Testing:
- [x] **Test cases** covering new functionality
- [x] **Integration tests** for agent interface compatibility  
- [x] **Breaking change validation** for all modified tools
- [x] **Model provider mocking** for reliable CI/CD
- [x] **Edge case validation** for error scenarios

## Checklist

- [x] I have read the CONTRIBUTING document
- [x] I have added tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly  
- [x] I have added an appropriate example to the documentation to outline the feature
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

- By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### **For Graph Tool Users:**
1. Change `agent_graph` to `graph` 
2. Restructure topology with `entry_points` array
3. Add `role` field to all nodes
4. Optionally add per-node model configuration

### **For use_llm Users:**
1. Change `use_llm` to `use_agent`
2. Optionally add `model_provider` for model switching
3. Optionally use `tools` parameter for selective access

### **For Swarm Users:**
1. Replace `swarm_size` with `agents` array
2. Define custom agent specifications with individual system prompts
3. Optionally configure per-agent tools and models
4. Remove `coordination_pattern` parameter

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -24,7 +24,8 @@ classifiers = [
     "Topic :: Software Development :: Libraries :: Python Modules",
 ]
 dependencies = [
-    "strands-agents>=0.1.0,<1.0.0",
+    "strands-agents @ git+https://github.com/strands-agents/sdk-python.git",
```
Comment (cagataycali) [inline pyproject.toml]: Currently dependency check fails due:

```
  Downloading a2a_sdk-0.2.11-py3-none-any.whl.metadata (4.8 kB)
ERROR: Cannot install a2a-sdk because these package versions have conflicting dependencies.

The conflict is caused by:
    a2a-sdk 0.2.12 depends on grpcio-tools>=1.60
    a2a-sdk 0.2.11 depends on grpcio-tools>=1.60
```

Dependencies will be updated after this change deployed to pypi: https://github.com/strands-agents/sdk-python/commit/1f64b4b819c412a684b677c17429604ecbcc66df

Comment (cagataycali) [reply pyproject.toml]: Updated the dependencies to use SDK directly, expected to fail in GitHub CI.

File: pyproject.toml [hunk 2]
```diff
@@ -25,6 +25,7 @@ classifiers = [
 ]
 dependencies = [
     "strands-agents>=0.1.0,<1.0.0",
+    "strands-agents[a2a]>=0.1.0,<1.0.0",
```
Comment (arronbailiss) [inline pyproject.toml]: won't need this import with v1

Comment (cagataycali) [reply pyproject.toml]: removed in next revision ^^

File: pyproject.toml (hunk 1 above)
Comment (cagataycali) [reply pyproject.toml]: `test_swarm.py` is failing due to changes not published into pypi yet, will re-run CI after pypi release.

File: src/strands_tools/graph.py [hunk 3]
```diff
@@ -0,0 +1,694 @@
...
+    3. Output from one agent propagates as input to dependent agents
+    4. Supports conditional routing and parallel execution where possible
+    5. Each agent can use different model providers and configurations
+
+    Key Differences from Old agent_graph:
+    -----------------------------------
+    - **Execution Model**: Task execution vs persistent message-passing
+    - **Communication**: Output propagation vs real-time message queues
+    - **Lifecycle**: Task-based execution vs long-running agent networks
+    - **Architecture**: Uses SDK Graph classes vs custom implementation
```
Comment (Unshure) [inline src/strands_tools/graph.py:434]: Do we need to mention the old agent graph in the tool description?

File: src/strands_tools/graph.py [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+    # Get graph status
+    result = agent.tool.graph(action="status", graph_id="research_pipeline")
+
+    # List all graphs
+    result = agent.tool.graph(action="list")
+
+    # Delete a graph
+    result = agent.tool.graph(action="delete", graph_id="research_pipeline")
+
+    Notes:
```
Comment (Unshure) [inline src/strands_tools/graph.py:526]: Can we add a note that if not defined, all of the parent tools are given to the sub_agent

File: src/strands_tools/graph.py [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+                        "tools": list[str] (optional)
+                    }, ...
+                ],
+                "edges": [{"from": str, "to": str}, ...],
+                "entry_points": [str, ...] (optional, auto-detected if not provided)
+            }
+        task: Task to execute through the graph (required for execute action).
+        model_provider: Default model provider for all agents in the graph.
+            Individual nodes can override this with their own model_provider.
+            Options: "bedrock", "anthropic", "litellm", "ollama", "openai", etc.
```
Comment (Unshure) [inline src/strands_tools/graph.py:458]: Can we write all supported values instead of etc? The agent wont know what to include here.

File: src/strands_tools/graph.py [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+        model_settings: Default model configuration for all agents.
```
Comment (Unshure) [inline src/strands_tools/graph.py:459]: How does the agent know what the model_setting should be for different providers? Can they only update model_id with this setting?

File: src/strands_tools/graph.py [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+        if graph_id not in self.graphs:
+            return {"status": "error", "message": f"Graph {graph_id} not found"}
+
+        try:
+            graph_info = self.graphs[graph_id]
+            graph = graph_info["graph"]
+
+            # Execute the graph
+            start_time = time.time()
+            result = graph(task)
```
Comment (Unshure) [inline src/strands_tools/graph.py:275]: Is this calling the `@tool` decorated method? Is that expected?

Trying to understand what this is calling.

File: src/strands_tools/graph.py [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+                    "failed_nodes": result.failed_nodes,
+                    "results": results_text,
+                },
+            }
+
+        except Exception as e:
+            logger.error(f"Error executing graph {graph_id}: {str(e)}")
+            return {"status": "error", "message": f"Error executing graph: {str(e)}"}
+
+    def get_graph_status(self, graph_id: str) -> Dict:
```
Comment (Unshure) [inline src/strands_tools/graph.py:310]: Do we asynchronously start the graph? If we wait for the execution to finish, what is this method for?

File: src/strands_tools/graph.py (hunk 3 above)
Comment (cagataycali) [reply src/strands_tools/graph.py:434]: Old agent_graph still do exists, I am not sure we need here ^^

File: src/strands_tools/graph.py (hunk 5 above)
Comment (cagataycali) [reply src/strands_tools/graph.py:458]: We can definitely expand the docstring, is this blocker?

File: src/strands_tools/graph.py (hunk 6 above)
Comment (cagataycali) [reply src/strands_tools/graph.py:459]: Depends on heavily the parent agent prompt, parent agent may decide which config needs to be used

File: src/strands_tools/graph.py (hunk 7 above)
Comment (cagataycali) [reply src/strands_tools/graph.py:275]: Calling builder's graph directly, not the @tool decorated function here ^^

File: src/strands_tools/graph.py (hunk 8 above)
Comment (cagataycali) [reply src/strands_tools/graph.py:310]: Graph is sync ^^
